src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocumentFactory.cs

```diff
@@ -15,94 +14,30 @@ namespace Microsoft.VisualStudio.LanguageServerClient.Razor
-            _fileUriProvider = fileUriProvider;
-        }
-
-        private IContentType CSharpLSPContentType
-        {
-            get
-            {
-                if (_csharpLSPContentType == null)
-                {
-                    var registeredContentType = _contentTypeRegistry.GetContentType(RazorLSPConstants.CSharpLSPContentTypeName);
-                    _csharpLSPContentType = new RemoteContentDefinitionType(registeredContentType);
-                }
-
-                return _csharpLSPContentType;
-            }
         }
 
-        public override bool TryCreateFor(ITextBuffer hostDocumentBuffer, out VirtualDocument virtualDocument)
-        {
-            if (hostDocumentBuffer is null)
-            {
-                throw new ArgumentNullException(nameof(hostDocumentBuffer));
-            }
-
-            if (!hostDocumentBuffer.ContentType.IsOfType(RazorLSPConstants.RazorLSPContentTypeName))
-            {
-                // Another content type we don't care about.
-                virtualDocument = null;
-                return false;
-            }
-
-            var hostDocumentUri = _fileUriProvider.GetOrCreate(hostDocumentBuffer);
-
-            // Index.cshtml => Index.cshtml__virtual.cs
-            var virtualCSharpFilePath = hostDocumentUri.GetAbsoluteOrUNCPath() + RazorLSPConstants.VirtualCSharpFileNameSuffix;
-            var virtualCSharpUri = new Uri(virtualCSharpFilePath);
-
-            var csharpBuffer = _textBufferFactory.CreateTextBuffer();
-            _fileUriProvider.AddOrUpdate(csharpBuffer, virtualCSharpUri);
-            csharpBuffer.Properties.AddProperty(RazorLSPConstants.ContainedLanguageMarker, true);
-            csharpBuffer.Properties.AddProperty(LanguageClientConstants.ClientNamePropertyKey, "RazorCSharp");
-
-            // Create a text document to trigger the C# language server initialization.
-            _textDocumentFactory.CreateTextDocument(csharpBuffer, virtualCSharpFilePath);
-
-            csharpBuffer.ChangeContentType(CSharpLSPContentType, editTag: null);
-
-            virtualDocument = new CSharpVirtualDocument(virtualCSharpUri, csharpBuffer);
-            return true;
-        }
+        protected override string HostDocumentContentTypeName => RazorLSPConstants.RazorLSPContentTypeName;
+        protected override string LanguageContentTypeName => RazorLSPConstants.CSharpLSPContentTypeName;
+        protected override string LanguageFileNameSuffix => RazorLSPConstants.VirtualCSharpFileNameSuffix;
+        protected override IContentType ConvertToLSPContentType(IContentType registeredContentType) => new RemoteContentDefinitionType(registeredContentType);
+        protected override Dictionary<object, object> AdditionalLanguageBufferProperties => _additionalLanguageBufferProperties;
+        protected override VirtualDocument CreateVirtualDocument(Uri uri, ITextBuffer textBuffer) =>
+            new CSharpVirtualDocument(uri, textBuffer);
```

Comment:
nit: I wouldn't worry about too long of lines, i'd just throw it on the one above.

Reply:
Sounds good

---
In reply to: [451979214](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r451979214) [](ancestors = 451979214)